### PR TITLE
[Bug #1267] avoid open redirection in register URL

### LINKED ISSUE #1267: Open redirect in logout URL
<b>Comment by <a href='https://openhatch.org/people/brittag/'>brittag</a>:</b><dl><dd>We got a report via email (thanks @M7K911) that <a href="http://openhatch.org/account/logout/?">http://openhatch.org/account/logout/?</a>
next=http:///<a href="http://www.google.com">www.google.com</a> works as an open redirect, and it probably shouldn't. This seems 
especially interesting because <a href="http://openhatch.org/account/logout/?next=http://www.google.com">http://openhatch.org/account/logout/?next=http://www.google.com</a> 
(with two slashes instead of three) just shows Django's "logged out" page.
   
</dl></dd><hr/>

<b>Comment by <a href='https://openhatch.org/people/mattskone/'>mattskone</a>:</b><dl><dd>This pull request contains the proposed fix: <a href="https://github.com/openhatch/oh-mainline/pull/234">https://github.com/openhatch/oh-mainline/pull/234</a>
   
</dl></dd><hr/>

<b>Comment by <a href='https://openhatch.org/people/paulproteus/'>paulproteus</a>:</b><dl><dd>I +1'd this and merged it. I also just deployed.

Huge thanks to all.
   
</dl></dd><hr/>

<b>Comment by <a href='https://openhatch.org/people/MTK/'>MTK</a>:</b><dl><dd>Hi @M7K911 here i just review the patch it's nicely patched there is another
Open redirect on different page but on same parameter here is the P.O.C .

<a href="http://openhatch.org/openid/register/?next=http://evil.com">http://openhatch.org/openid/register/?next=http://evil.com</a>

in this case user need to be logged in in order to get successfully redirected.
   
</dl></dd><hr/>

Status: <b>reopened</b>
Nosy List: <b>  brittag, mattskone, paulproteus </b>
Priority: <b>bug</b>
Imported from roundup ID: <b>952</b> (<a href='http://roundup-archive.openhatch.org/bugs/issue952'>view archived page</a>)
Last modified: <b>2014-05-06.13:52:38</b>

### PR DESCRIPTION
**Pull request by vipul-sharma20**
This fixes #1267, open redirection in `openid/register/` . Earlier `openid/register/?next=http://example.com` redirected user to `example.com` now it redirects to OpenHatch root. Also included tests for the same.

### CODE REVIEW

File: mysite/account/tests.py [hunk 1]
```diff
@@ -451,3 +451,27 @@ def test(self):
 
         self.assertTrue(self.user_logged_in(client1.session))
         self.assertFalse(self.user_logged_in(client2.session))
+
+class UrlRedirect(WebTest):
+    fixtures = ['user-paulproteus']
```
Comment (willingc) [inline mysite/account/tests.py]: It would be nice to add a docstring for this class.

File: mysite/account/tests.py [hunk 2]
```diff
@@ -451,3 +451,27 @@ def test(self):
...
+
+    def test_appended_url_redirect(self):
+        client = Client()
```
Comment (willingc) [inline mysite/account/tests.py]: If you can add a docstring for the test function, that would be helpful for the next person working on the code.

File: mysite/account/views.py [hunk 3]
```diff
@@ -413,7 +413,7 @@ def invite_someone_do(request):
 # modified trivially in the POST handler.
 
 
-@django_authopenid.views.not_authenticated
+@authenticated
```
Comment (willingc) [inline mysite/account/views.py]: @paulproteus should this be `@django_authopenid.views.authenticated` or is `@authenticated` sufficient

File: mysite/account/views.py [hunk 4]
```diff
@@ -530,7 +530,6 @@ def register(request,
     }, context_instance=django_authopenid.views._build_context(
         request, extra_context=extra_context))
 
-# We use this "not_authenticated" wrapper so that if you *are* logged in,
+# We use this "authenticated" wrapper so that if you *are* logged in,
 # you go straight to the ?next= value.
-login = django_authopenid.views.not_authenticated(
-    django.contrib.auth.views.login)
+login = authenticated(django.contrib.auth.views.login)
```
Comment (willingc) [inline mysite/account/views.py]: Similar question as for line 416

File: mysite/account/views.py (hunk 3 above)
Comment (vipul-sharma20) [inline mysite/account/views.py]: @willingc that's what was causing the problem. There were some issues with `@django_authopenid.views.authenticated` that's why we were getting open redirect. It had to be replaced with new decorator

File: mysite/base/decorators.py [hunk 5]
```diff
@@ -213,3 +214,17 @@ def get(self):
             return x
 
     return property(get)
+
+def authenticated(func):
+    """ decorator that redirect user to next page if
+    he is already logged."""
```
Comment (willingc) [inline mysite/base/decorators.py]: Typically, we try to keep documentation gender-neutral. Perhaps change "he" to "the user"

**Comment by willingc** [comment]
@vipul-sharma20 Thank you for making the changes that I requested. Since @paulproteus is more familiar with the code on authentication and this particular issue, I would like him to review those portions of the PR. As for all the rest of the PR, I am an enthusiastic +1. :cake:

**Comment by phildini** [comment]
So the code this is replacing is here: https://github.com/openhatch/oh-mainline/blob/master/vendor/packages/django-authopenid/django_authopenid/views.py#L152

The only real difference I can see is that it's hardcoding the default redirect value to '/', and urlparsing 'next'. With that, I have some questions.

1) Why not keep it as a setting in the settings file?
2) Is the urlparsing part of the bug? why do we expect next to not have a protocol?

/cc @paulproteus

**Comment by paulproteus** [comment]
Hi @vipul-sharma20 & @phildini ,

Vipul, thanks for your tireless work on this. When I saw the implementation, I was torn -- on one hand, I wanted to merge your great work, but on the other hand, it struck me that this issue would result in adding a lot of code.

I talked with @phildini and we came up with the idea of deciding that this issue is out of scope for the project. So I merged his docs change doing that.

So for now, @willingc and @vipul-sharma20 , I will close this. Even though it's great code, I just wonder if the project would be better off with even less code. Thank you both for the work and the careful review. @vipul-sharma20 let me know if you're still interested in contributing to OpenHatch -- I'm happy to help you find things to do.

**Pull request closed without merging**